src/librustc_middle/ty/query/profiling_support.rs

```diff
@@ -154,6 +156,53 @@ impl SpecIntoSelfProfilingString for DefIndex {
     }
 }
 
+impl SpecIntoSelfProfilingString for LocalDefId {
+    fn spec_to_self_profile_string(
+        &self,
+        builder: &mut QueryKeyStringBuilder<'_, '_, '_>,
+    ) -> StringId {
+        builder.def_id_to_string_id(DefId { krate: LOCAL_CRATE, index: self.local_def_index })
+    }
+}
+
+impl<T: SpecIntoSelfProfilingString> SpecIntoSelfProfilingString for WithOptConstParam<T> {
+    fn spec_to_self_profile_string(
+        &self,
+        builder: &mut QueryKeyStringBuilder<'_, '_, '_>,
+    ) -> StringId {
+        // We print `WithOptConstParam` values as tuples to make them shorter
+        // and more readable, without losing information:
+        //
+        // "WithOptConstParam { did: foo::bar, const_param_did: Some(foo::baz) }"
+        // becomes "(foo::bar, foo::baz)" and
+        // "WithOptConstParam { did: foo::bar, const_param_did: None }"
+        // becomes "(foo::bar, _)".
```

Comment:
What does self profiling need the `WithOptConstParam` for?

As long as you don't use `WithOptConstParam` to directly interact with querysystem you can most probably just store them with `self.did.to_self_profile_string(builder)` :thinking: 

You probably want to uniquely identify query calls? Considering that `self.const_param_did` should be unique for each `DefId` (as we otherwise cause an ICE while building mir) you might also be able to store this as `(did, bool)` for some additional wins.

Reply:
> What does self profiling need the `WithOptConstParam` for?

Some queries (e.g. `mir_build`) take `WithOptConstParam` values as keys. Since self-profiling can record a string representation for each query key that gets used some of those keys will be stringified `WithOptConstParam` values. You can find the full list of queries (and their keys) in https://github.com/rust-lang/rust/blob/master/src/librustc_middle/query/mod.rs.

> You probably want to uniquely identify query calls? Considering that `self.const_param_did` should be unique for each `DefId` (as we otherwise cause an ICE while building mir) you might also be able to store this as `(did, bool)` for some additional wins.

I'm not quite sure I follow. You are saying that there is an invariant around `const_param_did` that makes sure that it is always the same (or `None`) for a given `did` value? 

In any case, I think it's a good idea to keep this as close to the original value as possible. The encoding format should make sure that there is no data redundancy (assuming that that DefId was already mentioned somewhere already).

Reply:
> Since self-profiling can record a string representation for each query key that gets used

That's what I wasn't sure about, thanks :+1:

> I'm not quite sure I follow. You are saying that there is an invariant around const_param_did that makes sure that it is always the same (or None) for a given did value?

yes.

> In any case, I think it's a good idea to keep this as close to the original value as possible. The encoding format should make sure that there is no data redundancy (assuming that that DefId was already mentioned somewhere already).

:+1: